appcypher-syntax.md

```diff
@@ -0,0 +1,947 @@
+
+    ```swift
+    let empty_tuple = ()
+    ```
+
+- Tuple indexing. Tuple index space starts at index 0.
+
+    ```rust
+    let book = ("Harry Potter", "J.K. Rowling", 1995)
+    let title = book.1 // Harry Potter
+    let author = book.2 // J.K. Rowling
+    let year = book.3 // 1995
+    ```
+
+- Tuple destructuring.
+
+    ```rust
+    let person = ("Sam", 50)
+    let (name, age) = person
+    ```
+
+- Tuple unpacking.
+
+    ```swift
+    func add(a, b) {
+        a + b
+    }
+
+    let arguments = (5, 6)
+    add(...arguments)
+    ```
+
+- Tuples have limitations because they are expected to be amenable to static analysis.
+
+    ```swift
+    var tup1 = (1, 2)
+    var tup2 = (5, 6)
+    ```
+
+    Directly or indirectly, a tuple can't be used to extend itself.
+
+    ```go
+    for i in range(x) {
+        tup2 = (...tup2, ...tup1) // Invalid!
+    }
+
+    for i in range(x) {
+        tup3 = (...tup2, 1)
+        tup2 = (tup3, tup1) // Invalid!
+    }
+    ```
+
+- The type of a tuple.
+
+    Since tuples can't change in size, a tuple has a fixed type. For example the tuple `(0xFF, "Hi")` has type `(Int, String)`. `("Hello", 5.0, "Hi")` has type `(String, F64, String)`
+
+## LISTS
+- List is a dynamic container for storing and accessing items.
+
+    ```swift
+    var unordered_list = [7, 3, 8, 5, 4, 0, 9, 1, 2, 6]
+    ```
+
+    Empty list.
+
+    ```swift
+    var empty_list = []
+    ```
+
+    Using List constructor.
+
+    ```swift
+    let students = List::new()
+    ```
+
+- Type annotation for lists.
+
+    ```swift
+    var names: [String] = []
+    ```
+
+- Ratio lists can have heterogenous elements.
+
+    ```swift
+    var stuffs = [500, "Steve", 1.0] // Inferred
+    var stuffs: [F64|String|Int] = [500, "Steve", 1.0] // Explicit typing
+    ```
+
+- List indexing. List index space starts at index 0.
+
+    ```swift
+    stuffs[0] // 500
+    ```
+
+- Slicing a list.
+
+    ```swift
+    contestants[2..7] // 3rd to the 7th index
+    contestants[2..] // 3rd to the last index
+    contestants[..7] // first to the 7th index
+    contestants[..] // first to the last index
+    ```
+
+    A slice can have a step value.
+
+    ```swift
+    let odds = numbers[1..2..30] // The step value is the `2` in the middle.
+    ```
+
+## RANGE
+- Range is an iterable that allows iterating through a range of values.
+
+    ```swift
+    let days = 1..31 // 1 to 30
+    ```
+
+- A range can have a step value.
+
+    ```swift
+    let evens = 2..2..20 // The step value is the `2` in the middle.
+    ```
+
+
+## STRINGS
+- Strings are a sequence of valid UTF-8 codepoints and they are delimited by double quotes `"`.
+
+    ```swift
+    let language = "Ratio"
+    ```
+
+- String interpolation is done with a pair of curly braces
+    ```js
+    var story = "{language} was started in the {year}"
+    ```
+
+- String operations
+
+    ```js
+    var concatenate = "ab" + "c" // "abc"
+    var multiply = "ab" * 2 // "abcabc"
+    ```
```

Comment:
String is not a primitive type. In fact, there is no notion of primitiveness in my proposal. 

There are 
- scalar types, which are machine-level types because they are represented at the machine level
- and there are high-level types, which build on scalar types

`+` and `*` are overloaded for the String type. String is a high-level type because it builds on scalar type. Implementation-wise, a  string type would look like this.

```rust
type String {
    length: Int, 
    capacity: Int,
    buffer: *U8,
}
```